django_log_formatter_asim/__init__.py

```diff
@@ -25,6 +28,40 @@ def get_log_dict_with_raw(self, log_dict):
 
         return copied_dict
 
+    def _get_container_id(self):
+        """
+        The dockerId (container Id) is available via the metadata endpoint. However, it looks like it is embedded in the
+        metadata URL e.g.:
+        ECS_CONTAINER_METADATA_URI=http://169.254.170.2/v3/709d1c10779d47b2a84db9eef2ebd041-0265927825
+        See: https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-metadata-endpoint-v4-response.html
+        """
+
+        try:
+            return os.environ["ECS_CONTAINER_METADATA_URI"].split("/")[-1]
+        except (KeyError, IndexError):
+            return ""
+
+    def _datadog_trace_dict(self):
+        event_dict = {}
+
+        span = tracer.current_span()
+        trace_id, span_id = (
+            (str((1 << 64) - 1 & span.trace_id), span.span_id) if span else (None, None)
```

Comment:
I'm not familiar with this syntax.

In human speak, what is `(str((1 << 64) - 1 & span.trace_id)` doing?

Reply:
In all honesty, I wasn't completely sure what this was doing.


I was using the following snippet from the DataDog documentation: <https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/python/#no-standard-library-logging>

Reading up on what it's doing, `(1 << 64) - 1` returns the largest 64 bit integer - `(2 ** 64) - 1` might be a more understandable way to write this - the `& span.trace_id` runs an AND operation against the two numbers, which pulls out the first 64 bits of the `span.trace_id`.

On reflection, this is probably adding in uneccesarry complexity if we stick to using 64 bit trace ids but if we ever were to use 128 bit trace ids this would actually do something.

Here's an example of how this is working (and how it would work with 128 bit trace ids):

```python
>>> import random
>>> random_64_bit_int = random.getrandbits(64)
>>> random_64_bit_int
8134199743828943348
>>> trace_id = (1 << 64) - 1 & random_64_bit_int    # here random_64_bit_int is span.trace_id
>>> trace_id
8134199743828943348
>>>
>>>
>>> random_128_bit_int = random.getrandbits(128)
>>> random_128_bit_int
211872389340250636551607899401817332721
>>> trace_id = (1 << 64) - 1 & random_128_bit_int
>>> trace_id
1919324754494004209
>>> last_64_bits = bin(random_128_bit_int).strip('0b')[-64:]
>>> last_64_bits
'0001101010100010110011111011000111111000001110101001101111110001'
>>> int(last_64_bits, 2)
1919324754494004209
>>> int(last_64_bits, 2) == trace_id
True
```

Reply:
I wonder whether for readability and the next person to get here, we could extract a little method...

```python
def _first_64_bits_of(self, trace_id):
    # See https://docs.datadoghq.com/tracing/other_telemetry/connect_logs_and_traces/python/#no-standard-library-logging
    return (str((1 << 64) - 1 & trace_id)
```

...then use it here...
```suggestion
            self._first_64_bits_of(span.trace_id), span.span_id) if span else (None, None)
```